File: Python/src/Cracking the Coding Interview Problems/One_Five.py
"""1.5 Implement a method to perform basic string compression using the counts
of repeated characters. For example, the string aabcccccaaa would become
a2blc5a3. If the "compressed" string would not become smaller than the original
string, your method should return the original string."""


def main():
    s = raw_input("Please input a string to be compressed: ")
    s = compress_string(s)
    print s
    return


def compress_string(s):
    s_array = ['' for a in range(0, s.__len__())]
    hash_table = [0 for b in range(0, s.__len__())]
    count = 1
    index = 0
    temp = s[0]
    # write logic to get number of chars each
    for j in range(0, s.__len__()):
        if temp == s[j]:
            hash_table[index] += 1
            count += 1
            continue
        else:
            hash_table[index] = count
            index += 1
            count = 1
            temp = s[j]
    count = 0
    index = 0
    while count < s.__len__():
        s_array[index] = s[count]
        index += 1
        s_array[index] = str(hash_table[index - 1])
        index += 1
        count += hash_table[index - 2]
        print hash_table[index - 2]
    j = ""
    j = j.join(s_array)
    if j.__len__() >= s.__len__():
        return s
    else:
        return j


main()

File: Java/src/Cracking The Coding Interview/One_Five.java
package Ch1_InterviewQuestions;

import java.util.Scanner;

/*
 * 1.5 Implement a method to perform basic string compression using the counts
 * of repeated characters. For example, the string aabcccccaaa would become
 * a2blc5a3. If the "compressed" string would not become smaller than the original
 * string, your method should return the original string.
 */

public class One_Five {

	public static void main(String[] args) 
	{
		Scanner in = new Scanner(System.in);
		String OneFive = in.nextLine();
		System.out.println(compressString(OneFive));
	}
	
	static String compressString(String a)
	{
		int i = 0, count = 0;
		String temp = "";
		while(i < a.length())
		{
			if(a.charAt(i) == ' ')
			{
				temp += " ";
				if(i == a.length() - 1)
				{
					break;
				}
				i++;
				continue;
			}
			char current = a.charAt(i);
			while(i != a.length() && current == a.charAt(i))
			{
				count++;
				i++;
			}
			temp += current;
			temp += count;
			count = 0;
		}
		if(a.length() < temp.length())
		{
			return a;
		}
		a = temp;
		return a;
	}
	
}
